get-ghcup.sh

```diff
@@ -0,0 +1,228 @@
+# @USAGE: <no args>
+# @DESCRIPTION:
+# Checks to see if the ghcup executable exists.
+has_ghcup() {
+  command_exists "${GHCUP_EXE}"
+}
+
+# @FUNCTION: on_path
+# @USAGE: <path-piece>
+# @DESCRIPTION:
+# Checks to see if the <path-piece> is in the PATH environment variable.
+on_path() {
+  echo ":$PATH:" | grep -q :"$1":
+}
+
+# @FUNCTION: on_path
+# @USAGE: <no args>
+# @DESCRIPTION:
+# Checks to see if the ghcup bin directory is in the PATH environment variable.
+ghcup_on_path() {
+  on_path "${BIN_LOCATION}"
+}
+
+
+    ############################
+    #--[ Installation Logic ]--#
+    ############################
+
+
+# @FUNCTION: preamble
+# @USAGE: <no args>
+# @DESCRIPTION:
+# Prints a welcome message to the terminal and informs the user what the script
+# is going to do. They can cancel by pressing ctrl-c.
+preamble () {
+  echo
+  echo "Welcome to Haskell!"
+  echo
+  echo "This will download and install the Glasgow Haskell Compiler (GHC) for "
+  echo "the Haskell programming language, and the Cabal build tool."
+  echo
+  echo "It will add the 'cabal', 'ghc', and 'ghcup' executables to bin directory "
+  echo "located at: "
+  echo
+  echo "  ${BIN_LOCATION}"
+  echo
+  echo "This path will then be added to your PATH environment variable by "
+  echo "modifying the bashrc file at:"
+  echo
+  echo "  $HOME/.bashrc"
+  echo
+  echo "To proceed with the installation press enter, to cancel press ctrl-c"
+  echo
+
+  # Wait for user input to continue.
+  read
+}
+
+# @FUNCTION: check_for_ghcup_installation
+# @USAGE: <no args>
+# @DESCRIPTION:
+# Check if the 'ghcup' command already exists, if so provide instructions
+# for upgrading.
+check_for_ghcup_installation() {
+  if has_ghcup; then
+    echo
+    echo "Welcome to Haskell!"
+    echo
+    echo  "We noticed that you already have 'ghcup' installed:"
+    echo
+    echo "  Location: ${GHCUP_EXE}"
+    echo "  Version: `${GHCUP_EXE} --version`"
+    echo
+    echo "To upgrade to the latest version of ghcup run:"
+    echo
+    echo "  ghcup self-update"
+    echo
+    echo "To upgrade to the newest ghc and cabal versions run:"
+    echo
+    echo "  ghcup install && ghcup install-cabal"
+    echo
+
+    if ! ghcup_on_path; then
+      echo
+      echo "We noticed you don't have ghcup's bin directory ($HOME/.ghcup/bin)"
+      echo "in your PATH environment variable. To configure your current shell: "
+      echo
+      echo "  source $HOME/.ghcup/env"
+      echo
+    fi
+
+    exit 0
+  fi
+}
+
+# @FUNCTION: source_ghcup_env_from_bashrc
+# @USAGE: <no args>
+# @DESCRIPTION:
+# Check to see if the ghcup env file is sourced from .bashrc, if not add it.
+source_ghcup_env_from_bashrc () {
+  # Check to see if the .bashrc file sources the ghcup environement file, if
+  # not then add it.
+  if [ -z "`cat "$HOME/.bashrc" | grep 'source "$HOME/.ghcup/env"'`" ]; then
+    echo "Updating $HOME/.bashrc to source ghcup environment file..."
+    echo 'source "$HOME/.ghcup/env"' >> $HOME/.bashrc
```

Comment:
Uh, no, never :)

Reply:
Yeah I wasn't sold on this either, the rustup script does something similar, but adds it to .profile which is arguably the correct place. I used .bashrc to mirror the readme instructions but you're correct this could very well break things. It would be nice if system configuration was a bit more structured :)